### PR TITLE
consensus: Increase iterations to 255

### LINKED ISSUE #1174: consensus: Increase iterations to 255
#### Summary

As previously agreed, iterations limit should be increased to 255.

#### Possible solution design or implementation

Remove `CONSENSUS_MAX_STEP` and set `CONSENSUS_MAX_ITER` to 255.
Step number is not unique throughout the whole round but it's relative to the iteration number (e.g., it won't be Step 5, but Iteration 1, Step 2).

A pre-requisite for this change is replace `step` with `iteration` in the message header.

### PR DESCRIPTION
**Pull request by herr-seppia**
Resolves #1174

### CODE REVIEW

File: consensus/src/execution_ctx.rs [hunk 1]
```diff
@@ -225,24 +226,31 @@ impl<'a, DB: Database, T: Operations + 'static> ExecutionCtx<'a, DB, T> {
...
             round_update,
+            iteration,
             step,
             _executor: executor,
             sv_registry,
             quorum_sender,
         }
     }
 
+    pub fn total_step(&self) -> u16 {
```
Comment (goshawk-3) [inline consensus/src/execution_ctx.rs:237]: total_step sounds misleading as we  have `fn get_step(&self) -> u16 ` in MessageTrait that returns the absolute step number too. 

Maybe we can use a naming convention:
`get_step()/step()` ->  returns absolute/total step number in all traits/structs
`get_step_name()/step_name()` ->  returns step name only

step_name: StepName
step: u8

What do you think?

**Review by goshawk-3 — CHANGES_REQUESTED** [review]

File: consensus/src/ratification/handler.rs [hunk 2]
```diff
@@ -64,16 +64,17 @@ impl MsgHandler<Message> for RatificationHandler {
         &mut self,
         msg: Message,
         ru: &RoundUpdate,
-        step: u8,
+        iteration: u8,
+        _step: StepName,
         committee: &Committee,
     ) -> Result<HandleMsgOutput, ConsensusError> {
-        if step != self.curr_step {
+        if iteration != self.curr_iteration {
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply consensus/src/ratification/handler.rs:71]: StepName is not used anymore, we use iteration.
The consistency_check is done by `Self::unwrap_msg`.

File: consensus/src/validation/handler.rs [hunk 3]
```diff
@@ -100,16 +101,17 @@ impl MsgHandler<Message> for ValidationHandler {
         &mut self,
         msg: Message,
         _ru: &RoundUpdate,
-        step: u8,
+        iteration: u8,
+        _step: StepName,
         committee: &Committee,
     ) -> Result<HandleMsgOutput, ConsensusError> {
-        if step != self.curr_step {
+        if iteration != self.curr_iteration {
```
[the remark this answers is not in the document]
Comment (herr-seppia) [reply consensus/src/validation/handler.rs:108]: Same as above

File: consensus/src/ratification/handler.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (goshawk-3) [reply consensus/src/ratification/handler.rs:71]: Could anyone call `collect` with `iteration == self.curr_iteration` and with step=`StepName::Validation`?

File: consensus/src/execution_ctx.rs (hunk 1 above)
Comment (herr-seppia) [reply consensus/src/execution_ctx.rs:237]: Good idea!

File: consensus/src/ratification/handler.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (herr-seppia) [reply consensus/src/ratification/handler.rs:70]: I'm going to remove completely the `step` from `Collect`

[the remark this answers is not in the document]
Comment (herr-seppia) [reply consensus/src/ratification/handler.rs:70]: Thinking about it, the iteration is not need anymore neither.

The iteration is inside the message header, we can get rid of that

[the remark this answers is not in the document]
Comment (goshawk-3) [reply consensus/src/ratification/handler.rs:70]: > Thinking about it, the iteration is not need anymore neither.
> The iteration is inside the message header, we can get rid of that

Are you sure? Iteration is the current iteration a step is running while Message::Header::iteration can be from future or past. This is used when we are processing messages from past.

File: consensus/src/consensus.rs [hunk 4]
```diff
@@ -201,29 +200,25 @@ impl<T: Operations + 'static, D: Database + 'static> Consensus<T, D> {
...
-
-                    let step = iteration_counter.step_from_pos(pos);
-                    let name = phase.name();
-
+                for phase in phases.iter_mut() {
+                    let step_name = phase.to_step_name();
                     // Initialize new phase with message returned by previous
                     // phase.
-                    phase.reinitialize(&msg, ru.round, step).await;
+                    phase.reinitialize(&msg, ru.round, iter).await;
```
Comment (fed-franz) [inline consensus/src/consensus.rs:221]: Not sure it makes sense to call this `reinitialize` anymore. It should probably be `initialize`

File: consensus/src/ratification/handler.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (herr-seppia) [reply consensus/src/ratification/handler.rs:70]: If future the message is queued
If from past, there is `collect_from_past`

I've only removed `iteration` from "normal" collect

[the remark this answers is not in the document]
Comment (goshawk-3) [reply consensus/src/ratification/handler.rs:70]: The check is redundant then. I added it to make sure that noone would call `collect` with msg from the past.

File: consensus/src/consensus.rs (hunk 4 above)
Comment (herr-seppia) [reply consensus/src/consensus.rs:221]: Can we agree that is not on the scope of this PR?
No problem in change it, but I would like to first agree on what the name should be (if)

Comment (fed-franz) [reply consensus/src/consensus.rs:221]: Sure, it's not part of this PR, that's why I just added a comment.
I think `initialize` is fine, but if you have something more aligned to standard (or our) naming convention, go for it

**Review by goshawk-3 — APPROVED** [review]

**Review by fed-franz — APPROVED** [review]
LGTM

**Pull request merged**